File: Java/Guppy.java
public class Guppy extends Fish {

  private int growthStep;
  private int foodCapacity;
  private int produceTime;
  private int growthTimer;
  private int produceTimer;

  /**
   * Konstruktor default guppy.
   */
  public Guppy() {
    maxStarving = 500;
    maxStarvationPeriod = 200;
    maxMoveTime = 70;
    growthTimer = 6;
    produceTimer = 60;
    id = counter++;
    speed = 1;
    harga = 50;
    growthStep = 1;
    foodCapacity = 0;
    rightDirect = false;
    degree = 180;
    radius = 20;
    jenis = "Guppy";
    setStarvationPeriod(maxStarvationPeriod);
    setStarving(maxStarving);
    setMoveTime(maxMoveTime);
    setProduceTime(getMaxProduceTime());
  }

  public boolean chaseFood(double radMin,Point pointKejar){
    if (radMin <= getRadius()) {
      System.out.println("REALLY HUNGRY :((");
      return true;
    } else {
      swimto(pointKejar, getSpeed());
      setStarving(getStarving() - 1);
      return false;
    }
  }

  public void biteFood(Aquarium aquarium,int idx){
    foodCapacity = foodCapacity + 1;
    if (foodCapacity >= growthTimer) { //next growth
      foodCapacity = 0;
      if (growthStep < 3) {
        growthStep++;
        this.harga += harga;
      }
    }
    setStarvationPeriod(maxStarvationPeriod);
    setStarving(maxStarving);
    aquarium.removeObject(aquarium.getListObjekMati().get(idx));
}
  /**
   * Makan.
   * @param aquarium Aquarium ikan
   */
  public void eat(Aquarium aquarium) {
    boolean existFood = false, eatFood = false;
    double radMin = 1e7; //untuk nyimpen jarak terdekat
    int i = 0, idx = 0;
    Point idxGuppy = getPosisi();
    Point pointKejar = new Point();
    while (i < aquarium.getListObjekMati().totalElmt()) {
      if (aquarium.getListObjekMati().get(i).getJenis().equals("Makanan Ikan")) {
        existFood = true;
        Point pointMakanan = aquarium.getListObjekMati().get(i).getPosisi();
        double jarak = idxGuppy.hitungJarak(pointMakanan);
        if (jarak < radMin) {
          radMin = jarak;
          idx = i;
          pointKejar = pointMakanan;
        }
      }
      i++;
    }

    if (existFood) {
      eatFood = chaseFood(radMin,pointKejar);
    }

    if (eatFood) {
      System.out.println("I'M HUNGRY :(");
      biteFood(aquarium,idx);
    } else {
      setStarving(getStarving() - 1);
    }
  }

  public void produce(Aquarium aquarium) {
    Point posisiKoin = new Point(getPosisi().getAbsis(), getPosisi().getOrdinat());
    Koin koin = new Koin(50 * growthStep, posisiKoin);
    aquarium.addObject(koin);
  }

  //Growing
  public void nextGrowth() {
    growthStep = growthStep + 1;
  }

  //Getter Setter
  public int getGrowthStep() {
    return growthStep;
  }

  public int getFoodCapaty() {
    return foodCapacity;
  }

  public int getProduceTime() {
    return produceTime;
  }

  public int getId() {
    return id;
  }


  public int getMaxProduceTime() {
    return produceTimer;
  }

  public void setFoodCapacity(int x) {
    foodCapacity = x;
  }

  public void setProduceTime(int x) {
    produceTime = x;
  }

}

File: C++/Guppy.cpp
#include "Guppy.hpp"
#include "Aquarium.hpp"

Guppy :: Guppy() : growth_timer(6), produce_timer(100){
	MAX_STARVING = 500;
	MAX_STARVATION_PERIOD = 200;
	MAX_MOVE_TIME = 50;
	speed  = 1;
	harga = 50;
	growth_step = 1;
	food_capacity = 0;
	right_direct = false;
	degrees = 90;
	radius = 20;
	jenis = "Guppy";
	setStarvationPeriod(MAX_STARVATION_PERIOD);
	setStarving(MAX_STARVING);
	setMoveTime(MAX_MOVE_TIME);
	setProduceTime(getMaxProduceTime());
}

void Guppy :: eat(Aquarium &aq){
	bool existFood = false;
	bool eat_food = false;
	double rad_min = 1e7; //untuk nyimpen jarak terdekat
  	int i=0,idx;
  	Point idx_guppy = getPosisi();
  	Point point_kejar;
  	while (i < aq.getListObjekMati().totalElmt()){
	    if (aq.getListObjekMati().get(i)->getJenis() == "Makanan Ikan"){
	    	existFood = true;
	    	Point point_makanan = aq.getListObjekMati().get(i)->getPosisi();
	    	double jarak = idx_guppy.hitungJarak(point_makanan);
	    	if(jarak < rad_min){
	    		rad_min = jarak;
	    		idx = i;
	    		point_kejar = point_makanan;
	    	}
	    }
	    i++;
  	}
  	// Cek mau dimakan atau dikejar
  	if(existFood){
  		if(rad_min <= getRadius()){
  			eat_food = true;
  		}
  		else{
  			swimto(point_kejar, getSpeed());
				setStarving(getStarving()-1);
  		}
  	}

	// Makan makanan
	if(eat_food){
		//printf("makan\n");

		food_capacity = food_capacity + 1;
		//If its a time to growth
		if(food_capacity >= growth_timer){ //next growth
			food_capacity = 0;
			if(growth_step<3){
				growth_step++;
				harga+=harga;
			}
		}
		//Update hungry time
		//real_eat
		setStarvationPeriod(MAX_STARVATION_PERIOD);
		setStarving(MAX_STARVING);
		//ilangin makanan ikan
		aq.RemoveObject(aq.getListObjekMati().get(idx));
	}
	else{
		setStarving(getStarving()-1);
	}
}

void Guppy :: produce(Aquarium& aq){
	Koin *koin = new Koin(50*growth_step,getPosisi());
	aq.AddObject(koin);
}

//Growing
void Guppy :: Next_Growth(){
	growth_step = growth_step + 1;
}

//Getter Setter
int Guppy :: getGrowthStep(){
	return growth_step;
}
int Guppy :: getFoodCapaty(){
	return food_capacity;
}
int Guppy :: getProduceTime(){
	return produce_time;
}

int Guppy :: getMaxProduceTime(){
	return produce_timer;
}

void Guppy :: setFoodCapacity(int x){
	food_capacity = x;
}

void Guppy :: setProduceTime(int x){
	produce_time = x;
}
